internal/controller/controlplaneauth/controller.go

```diff
@@ -0,0 +1,214 @@
+
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1alpha1.ControlPlaneAuthGroupVersionKind),
+		managed.WithExternalConnecter(&connector{
+			kube:  mgr.GetClient(),
+			usage: resource.NewProviderConfigUsageTracker(mgr.GetClient(), &apisv1alpha1.ProviderConfigUsage{}),
+		}),
+		managed.WithLogger(o.Logger.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
+		managed.WithInitializers(),
+		managed.WithConnectionPublishers(cps...))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(o.ForControllerRuntime()).
+		For(&v1alpha1.ControlPlaneAuth{}).
+		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
+}
+
+// A connector is expected to produce an ExternalClient when its Connect method
+// is called.
+type connector struct {
+	kube  client.Client
+	usage resource.Tracker
+}
+
+// Connect typically produces an ExternalClient by:
+// 1. Tracking that the managed resource is using a ProviderConfig.
+// 2. Getting the managed resource's ProviderConfig.
+// 3. Getting the credentials specified by the ProviderConfig.
+// 4. Using the credentials to form a client.
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.ControlPlaneAuth)
+	if !ok {
+		return nil, errors.New(errNotControlPlaneAuth)
+	}
+
+	if err := c.usage.Track(ctx, mg); err != nil {
+		return nil, errors.Wrap(err, errTrackPCUsage)
+	}
+
+	cfg, profile, err := upclient.NewConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+
+	return &external{
+		tokens:   tokens.NewClient(cfg),
+		accounts: accounts.NewClient(cfg),
+		profile:  profile,
+		kube:     c.kube,
+	}, nil
+}
+
+// An ExternalClient observes, then either creates, updates, or deletes an
+// external resource to ensure it reflects the managed resource's desired state.
+type external struct {
+	kube     client.Client
+	accounts *accounts.Client
+	profile  upclient.Profile
+	tokens   *tokens.Client
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ControlPlaneAuth)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotControlPlaneAuth)
+	}
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{}, nil
+	}
+	uid, err := uuid.Parse(meta.GetExternalName(cr))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, fmt.Sprintf("failed to parse external name as UUID %s", meta.GetExternalName(cr)))
+	}
+	resp, err := c.tokens.Get(ctx, uid)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(uperrors.IsNotFound, err), "failed to get token")
+	}
+
+	cr.Status.SetConditions(v1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: resp.AttributeSet["name"] == cr.Spec.ForProvider.ControlPlaneName,
```

Comment:
`Update` is empty anyway, so we might as well return `true` here. I think the only case we may need an update is if the token expires, and that's not possible today I believe, right?